src/ccs.rs

```diff
@@ -0,0 +1,453 @@
+  pub(crate) c: Vec<usize>,
+}
+
+/// A type that holds a witness for a given CCS instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct CCSWitness<G: Group> {
+  W: Vec<G::Scalar>,
+}
+
+// TODO: Make sure this is in the right form for committed CCS using MLE, possibly a separate type?
+/// A type that holds an CCS instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct CCSInstance<G: Group> {
+  pub(crate) comm_W: Commitment<G>,
+  pub(crate) X: Vec<G::Scalar>,
+}
+
+impl<G: Group> CCS<G> {
+  // TODO: Update commitment_key variables here? This is currently based on R1CS with M length
+  /// Samples public parameters for the specified number of constraints and variables in an CCS
+  pub fn commitment_key(S: &CCSShape<G>) -> CommitmentKey<G> {
+    let num_cons = S.num_cons;
+    let num_vars = S.num_vars;
+    let total_nz = S.M.iter().fold(0, |acc, m| acc + m.len());
+
+    G::CE::setup(b"ck", max(max(num_cons, num_vars), total_nz))
+  }
+}
+
+impl<G: Group> CCSShape<G> {
+  /// Create an object of type `CCSSShape` from the explicitly specified CCS matrices
+  pub fn new(
+    num_cons: usize,
+    num_vars: usize,
+    num_io: usize,
+    M: &[Vec<(usize, usize, G::Scalar)>],
+    t: usize,
+    q: usize,
+    d: usize,
+    S: Vec<Vec<usize>>,
+    c: Vec<usize>,
+  ) -> Result<CCSShape<G>, NovaError> {
+    let is_valid = |num_cons: usize,
+                    num_vars: usize,
+                    num_io: usize,
+                    matrix: &[(usize, usize, G::Scalar)]|
+     -> Result<(), NovaError> {
+      let res = (0..matrix.len())
+        .map(|i| {
+          let (row, col, _val) = matrix[i];
+          if row >= num_cons || col > num_io + num_vars {
+            Err(NovaError::InvalidIndex)
+          } else {
+            Ok(())
+          }
+        })
+        .collect::<Result<Vec<()>, NovaError>>();
+
+      if res.is_err() {
+        Err(NovaError::InvalidIndex)
+      } else {
+        Ok(())
+      }
+    };
+
+    // Check that the row and column indexes are within the range of the number of constraints and variables
+    let res_M = M
+      .iter()
+      .map(|m| is_valid(num_cons, num_vars, num_io, m))
+      .collect::<Result<Vec<()>, NovaError>>();
+
+    // If any of the matricies are invalid, return an error
+    if res_M.is_err() {
+      return Err(NovaError::InvalidIndex);
+    }
+
+    // We require the number of public inputs/outputs to be even
+    if num_io % 2 != 0 {
+      return Err(NovaError::OddInputLength);
+    }
+
+    let shape = CCSShape {
+      num_cons,
+      num_vars,
+      num_io,
+      M: M.to_vec(),
+      t,
+      q,
+      d,
+      S,
+      c,
+    };
+
+    Ok(shape)
+  }
+
+  // NOTE: Not using previous used multiply_vec (r1cs.rs), see utils.rs
+
+  // NOTE: Equivalent to is_sat_relaxed (r1cs.rs) but for CCCSS/LCCCS?
+  // Either here or as a separate method on LCCCS struct
```

Comment:
Since we don't quite need a relaxed variant of CCCS, we could have a separate type for LCCCS and an associated is_sat method.